src/libraries/System.Formats.Tar/src/System/Formats/Tar/TarWriter.Unix.cs

```diff
@@ -0,0 +1,96 @@
+// Licensed to the .NET Foundation under one or more agreements.
+// The .NET Foundation licenses this file to you under the MIT license.
+
+using System.Diagnostics;
+using System.IO;
+
+namespace System.Formats.Tar
+{
+    // Unix specific methods for the TarWriter class.
+    public sealed partial class TarWriter : IDisposable, IAsyncDisposable
+    {
+        // Unix specific implementation of the method that reads an entry from disk and writes it into the archive stream.
+        partial void ReadFileFromDiskAndWriteToArchiveStreamAsEntry(string fullPath, string entryName)
+        {
+            Interop.Sys.FileStatus status = default;
+            status.Mode = default;
+            status.Dev = default;
+            Interop.CheckIo(Interop.Sys.LStat(fullPath, out status));
+
+            TarEntryType entryType = (status.Mode & (uint)Interop.Sys.FileTypes.S_IFMT) switch
+            {
+                // Hard links are treated as regular files.
+                // Unix socket files do not get added to tar files.
+                Interop.Sys.FileTypes.S_IFBLK => TarEntryType.BlockDevice,
+                Interop.Sys.FileTypes.S_IFCHR => TarEntryType.CharacterDevice,
+                Interop.Sys.FileTypes.S_IFIFO => TarEntryType.Fifo,
+                Interop.Sys.FileTypes.S_IFLNK => TarEntryType.SymbolicLink,
+                Interop.Sys.FileTypes.S_IFREG => Format is TarFormat.V7 ? TarEntryType.V7RegularFile : TarEntryType.RegularFile,
+                Interop.Sys.FileTypes.S_IFDIR => TarEntryType.Directory,
+                _ => throw new IOException(string.Format(SR.TarUnsupportedFile, fullPath)),
+            };
+
+            FileSystemInfo info = entryType is TarEntryType.Directory ? new DirectoryInfo(fullPath) : new FileInfo(fullPath);
+
+            TarEntry entry = Format switch
+            {
+                TarFormat.V7 => new V7TarEntry(entryType, entryName),
+                TarFormat.Ustar => new UstarTarEntry(entryType, entryName),
+                TarFormat.Pax => new PaxTarEntry(entryType, entryName),
+                TarFormat.Gnu => new GnuTarEntry(entryType, entryName),
+                _ => throw new FormatException(string.Format(SR.TarInvalidFormat, Format)),
+            };
+
+            if ((entryType is TarEntryType.BlockDevice or TarEntryType.CharacterDevice) && status.Dev > 0)
+            {
+                uint major;
+                uint minor;
+                unsafe
+                {
+                    Interop.CheckIo(Interop.Sys.GetDeviceIdentifiers((ulong)status.Dev, &major, &minor));
```

Comment:
@tmds I'm seeing a weird error when retrieving the major and minor of a device file: The numbers do not match the expected values when retrieving them with the new syscalls.

Unfortunately, I'm struggling to find any C usage examples of `major()` and `minor()`, so I was hoping you could help  me.

Do you know if the `Dev` field of the `FileStatus` struct is the correct `dev_t` object I need to pass to this p/invoke, or am I confusing the `dev_t` object with another one?

Do you know if the `major()` and `minor()` syscalls should receive a different value than the one I'm currently passing in `pal_io.c`?


Here's a test that can directly repro the issue. Fetch my branch, then paste this code in any of my test files (the test csproj is already importing the interop files):

```cs
[Fact]
public void Test()
{
    // Before running this, create this block device <replace your username>:
    //      sudo mknod /home/username/myblock b 4 5
    // Then double check with a 'ls -l /home/username/' that the expected devMajor and devMinor showed up (4 and 5)

    uint expectedMajor = 4;
    uint expectedMinor = 5;
    string path = "/home/username/myblock";

    Interop.Sys.FileStatus status = default;
    status.Mode = default;
    status.Dev = default;
    Interop.CheckIo(Interop.Sys.LStat(fullPath, out status));

    uint major;
    uint minor;
    unsafe
    {
        Interop.CheckIo(Interop.Sys.GetDeviceIdentifiers((ulong)status.Dev, &major, &minor));
    }
    // Both of these checks fail
    Assert.Equal(expectedMajor, major);
    Assert.Equal(expectedMinor, minor);
}
```

Reply:
You need to use the `st_rdev` field, instead of the `st_dev` field.